File: 2term/2.3. hash/2.3. hash/Hash.cs
using ListNamespace;

namespace HashNamespace
{
    /// <summary>
    /// Hash.
    /// </summary>
    public class Hash  
    {
        const int size = 100;
        
        /// <summary>
        /// Elements of a hashtable are lists.
        /// </summary>
        private List[] bucket = new List[size];

        /// <summary>
        /// Create internal lists ih a hashtable.
        /// </summary>
        public Hash()
        {
            for (int i = 0; i < size; ++i)
            {
                bucket[i] = new List();
            }
        }

        /// <summary>
        /// Try to create a random number for indexing fields of hashtable.
        /// </summary>
        /// <param name="newElement"></param>
        /// <returns></returns>
        private int HashFunction(int newElement)
        {
            int sum = 0;
            for (int i = 0; i < newElement; ++i)
            {
                sum = sum + i * i;
            }
            sum = sum % 100;
            return sum;
        }

        /// <summary>
        /// Insert value in hashtable.
        /// </summary>
        /// <param name="newElement"></param>
        public void InsertToHashTable(int newElement)
        {
            bucket[HashFunction(newElement)].Push(newElement);
        }

        /// <summary>
        /// Check of element's exictence in a hashtable. 
        /// </summary>
        /// <param name="newElement"></param>
        /// <returns></returns>
        public bool Exist(int newElement)
        {
            return bucket[HashFunction(newElement)].Exist(newElement);
        }

        /// <summary>
        /// Analogue of Exist method but it is more comfortable for using in Main.cs.
        /// </summary>
        /// <param name="newElement"></param>
        public void ExistenceInPrinting(int newElement)
        {
            bool result = Exist(newElement);
            if (result)
            {
                System.Console.WriteLine("{0} is here!", newElement);
            }
            else
            {
                System.Console.WriteLine("{0} is not here!", newElement);
            }
        }

        /// <summary>
        /// Remove element from hashtable.
        /// </summary>
        /// <param name="newElement"></param>
        public void Remove(int newElement)
        {
            bucket[HashFunction(newElement)].RemovingByPosition(bucket[HashFunction(newElement)].GettingOfPositionByValue(newElement));
        }

        /// <summary>
        /// Print hashtable.
        /// </summary>
        public void PrintHashTable()
        {
            string result = "";
            for (int i = 0; i < size; ++i)
            {
                result = bucket[i].Print();
                if (result != "-")
                {
                    System.Console.WriteLine("{0}", bucket[i].Print());
                }
            }
        }

        /// <summary>
        /// Delete a hashtable.
        /// </summary>
        public void DeleteHashTable()
        {
            for (int i = 0; i < size; ++i)
            {
                bucket[i].Removing();
            }
        }
    }
}

File: 1term/8.1/классная работа от 13.11.2013/классная работа от 13.11.2013/hash.cpp
#include "hash.h"
#include "list.h"
#include <string.h>

const int b = 10;
using namespace list;

struct hash::HashTable
{
	List *bucket[b];
};
hash::HashTable *hash::createHashTable()
{
	HashTable *hashTable = new HashTable;
	for (int i = 0; i < b; ++i)
	{
		hashTable->bucket[i] = create();
		/*return hashTable;*/
	}
	return hashTable;
}

void hash::deleteHashTable(HashTable *hashTable)
{
	for (int i = 0; i < b; ++i)
	{
		removeList(hashTable->bucket[i]);
	}
	delete hashTable;
}

int hashFunction(ElementType newElement)
{
	int sum = 0;
	for (int i = 0; i < strlen(newElement); ++i)
	{
		int temp = (int) newElement[i];
		sum = sum + temp * i * i * i;
	}
	sum = sum % b;
	return sum;
}

void hash::insertToHashTable(HashTable *hashTable, ElementType newElement)
{
	insertToHead(hashTable->bucket[hashFunction(newElement)], newElement);
}

bool hash::exist(HashTable *hashTable, ElementType element)
{
	return exist(hashTable->bucket[hashFunction(element)], element);
}
void hash::remove(HashTable *hashTable, ElementType element)
{
	remove(hashTable->bucket[hashFunction(element)], positionOfElement(hashTable->bucket[hashFunction(element)], element));
}

void hash::printHashTable(HashTable *hashTable)
{
	for (int i = 0; i < b; ++i)
	{
		printOfTheList(hashTable->bucket[i]);
	}
}